shared_model/interfaces/iroha_internal/transaction_batch_template_definitions.hpp

```diff
@@ -0,0 +1,115 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef IROHA_TRANSACTION_BATCH_TEMPLATE_DEFINITIONS_HPP
+#define IROHA_TRANSACTION_BATCH_TEMPLATE_DEFINITIONS_HPP
+
+#include "interfaces/iroha_internal/transaction_batch.hpp"
+
+namespace shared_model {
+  namespace interface {
+
+    namespace {
+      /**
+       * check if all transactions belong to the same batch
+       * @param txs transactions to be checked
+       * @return true if all transactions from the same batch and false
+       * otherwise
+       */
+      bool allTxsInSameBatch(const types::SharedTxsCollectionType &txs) {
+        if (txs.size() == 1) {
+          return true;
+        }
+
+        // take batch meta of the first transaction and compare it with batch
+        // metas of remaining transactions
+        auto batch_meta = txs.front()->batchMeta();
+        if (not batch_meta) {
+          return false;
+        }
+
+        return std::none_of(++txs.begin(),
+                            txs.end(),
+                            [front_batch_meta = batch_meta.value()](
+                                const std::shared_ptr<Transaction> tx) {
+                              return tx->batchMeta()
+                                  ? **tx->batchMeta() != *front_batch_meta
+                                  : false;
+                            });
+      };
+    }  // namespace
+
+    template <typename TransactionValidator, typename FieldValidator>
+    iroha::expected::Result<TransactionBatch, std::string>
+    TransactionBatch::createTransactionBatch(
+        const types::SharedTxsCollectionType &transactions,
+        const validation::TransactionsCollectionValidator<TransactionValidator>
+            &validator,
+        const FieldValidator &field_validator) {
+      auto answer = validator.validate(transactions);
+
+      std::string reason_name = "Transaction batch: ";
+      validation::ReasonsGroupType batch_reason;
+      batch_reason.first = reason_name;
+      if (boost::empty(transactions)) {
+        batch_reason.second.emplace_back(
+            "Provided transactions are not from the same batch");
+      }
+      if (not allTxsInSameBatch(transactions)) {
+        batch_reason.second.emplace_back(
+            "Provided transactions are not from the same batch");
+      }
+      bool has_at_least_one_signature =
+          std::any_of(transactions.begin(),
+                      transactions.end(),
+                      [&field_validator, &batch_reason](const auto tx) {
+                        const auto &signatures = tx->signatures();
+                        if (not boost::empty(signatures)) {
+                          field_validator.validateSignatures(
+                              batch_reason, signatures, tx->payload());
+                          return true;
+                        }
+                        return false;
+                      });
+
+      if (not has_at_least_one_signature) {
```

Comment:
Seems that number of `if` branches can be reduced. Move all error cases to one branch.

Reply:
I optimised it a bit, but all if-s cannot be in one branch, because types of errors are different